tests/test_flow/test_persistence_random.py

```diff
@@ -0,0 +1,375 @@
+        @bn.version({binding.func_version})
+        @bn.persist({binding.should_persist})
+        def _({joined_dep_entity_names}):
+            record_call({binding.out_descriptor!r})
+            return compute_value([{joined_dep_entity_names}])
+        """
+        exec(dedent(raw_func_code), vars_dict)
+
+    def _compute_model_value(self, entity_name, memoized_values_by_entity_name=None):
+        if memoized_values_by_entity_name is None:
+            memoized_values_by_entity_name = {}
+        if entity_name in memoized_values_by_entity_name:
+            return memoized_values_by_entity_name[entity_name]
+
+        entity = self._entities_by_name[entity_name]
+        binding = entity.binding
+
+        if not binding.has_persisted_values:
+            dep_values = [
+                self._compute_model_value(dep_entity_name)
+                for dep_entity_name in binding.dep_entity_names
+            ]
+            values_by_entity_name = binding.compute_and_save_values(dep_values)
+            self._descriptors_computed_by_model.append(binding.out_descriptor)
+
+        else:
+            values_by_entity_name = binding.persisted_values_by_entity_name
+
+        memoized_values_by_entity_name.update(values_by_entity_name)
+        return values_by_entity_name[entity_name]
+
+    def _clear_called_descriptors(self):
+        self._descriptors_computed_by_flow[:] = []
+        self._descriptors_computed_by_model[:] = []
+
+    def _build_flow(self):
+        self.flow = self._builder.build()
+
+
+@attr.s
+class ModelEntity:
+    """Represents an entity in a Bionic flow."""
+
+    name = attr.ib()
+    should_persist = attr.ib()
+
+    binding = attr.ib(default=None)
+    dependent_entities = attr.ib(default=attr.Factory(list))
+
+    def compute_value(self, dep_values):
+        return f"{self.name}.{self.binding.func_version}({','.join(dep_values)})"
+
+
+@attr.s
+class ModelBinding:
+    """
+    Represents a user-provided function that computes the value of one or more entities.
+    """
+
+    out_dnode = attr.ib()
+    out_entities = attr.ib()
+    dep_entity_names = attr.ib()
+    func_version = attr.ib(default=1)
+
+    has_persisted_values = attr.ib(default=False)
+    persisted_values_by_entity_name = attr.ib(default=None)
+    some_descendant_might_have_persisted_values = attr.ib(default=False)
+
+    def update_func_version(self):
+        self.func_version += 1
+        self.invalidate_persisted_values()
+
+    def compute_value_for_outputs(self, dep_values):
+        return tuple(
+            out_entity.compute_value(dep_values) for out_entity in self.out_entities
+        )
+
+    def _compute_entity_value(self, entity_name, dep_values):
+        return f"{entity_name}.{self.func_version}({','.join(dep_values)})"
+
+    def compute_and_save_values(self, dep_values):
+        values_by_entity_name = {
+            entity.name: self._compute_entity_value(entity.name, dep_values)
```

Comment:
You can replace `self._compute_entity_value` with `entity.compute_value` here. Once you do, you can then remove `_compute_entity_value` method and avoid duplicating that logic in two places.

Reply:
Oh yeah, you're right. That was the result of an incomplete refactor.